dvc/cache/base.py

```diff
@@ -550,3 +555,70 @@ def get_files_number(self, path_info, hash_, filter_info):
             filter_info.isin_or_eq(path_info / entry[self.tree.PARAM_CHECKSUM])
             for entry in self.get_dir_cache(hash_)
         )
+
+    def _to_dict(self, dir_info):
+        return {
+            entry[self.tree.PARAM_RELPATH]: entry[self.tree.PARAM_CHECKSUM]
+            for entry in dir_info
+        }
+
+    @staticmethod
+    def _diff(ancestor, other):
+        from dictdiffer import diff
+
+        result = list(diff(ancestor, other))
+        for typ, _, _ in result:
+            if typ != "add":
+                raise MergeError(
+                    "unable to auto-merge directories with diff that contains "
+                    f"'{typ}'ed files"
+                )
+        return result
+
+    def _merge_dirs(self, ancestor_info, our_info, their_info):
+        from dictdiffer import patch
+        from operator import itemgetter
+
+        ancestor = self._to_dict(ancestor_info)
+        our = self._to_dict(our_info)
+        their = self._to_dict(their_info)
+
+        our_diff = self._diff(ancestor, our)
+        if not our_diff:
+            return their
+
+        their_diff = self._diff(ancestor, their)
+        if not their_diff:
+            return our
+
+        merged = patch(our_diff + their_diff, ancestor, in_place=True)
+
+        result = [
+            {
+                self.tree.PARAM_RELPATH: relpath,
+                self.tree.PARAM_CHECKSUM: checksum,
+            }
+            for relpath, checksum in merged.items()
+        ]
+
+        # Sorting the list by path to ensure reproducibility
+        return sorted(result, key=itemgetter(self.tree.PARAM_RELPATH))
```

Comment:
We have a similar thing in _collect_dir in BaseTree's, but these will be unified as a part of https://github.com/iterative/dvc/issues/4144 , once dir hash computation is detached from cache.